pkg/config/setup/config.go

```diff
@@ -1765,38 +1767,104 @@ func ResolveSecrets(config pkgconfigmodel.Config, secretResolver secrets.Compone
 			return fmt.Errorf("unable to marshal configuration to YAML to decrypt secrets: %v", err)
 		}
 
-		err = secretResolver.ResolveWithCallback(
-			yamlConf,
-			origin,
-			func(yamlPath []string, value any) bool {
-				settingName := strings.Join(yamlPath, ".")
-
-				// We received a notification about an unknown setting. This means that the a value
-				// inside a map was updated (ie: settings like additional_endpoints,
-				// kubernetes_node_annotations_as_tags, ...). This is an issue when a setting use a
-				// map[string] as a value. The key can contain a '.' making it impossible to set or get.
-				//
-				// The secrets resolver doesn't have the notion of what is a known setting and what
-				// isn't. Returning false tells the secretResolver that we refuse the notification and
-				// we want to be notified for the parent key.
-				//
-				// See secrets.ResolveCallback documentation for more information.
-				//
-				if !config.IsKnown(settingName) {
-					return false
-				}
-
-				log.Debugf("replacing handle for setting '%s' with secret value", settingName)
-				config.Set(settingName, value, pkgconfigmodel.SourceAgentRuntime)
-				return true
-			})
-		if err != nil {
+		secretResolver.Subscribe(func(handle string, settingPath []string, oldValue, newValue any) {
+			if err := configAssignAtPath(config, settingPath, newValue); err != nil {
+				log.Errorf("could not assign to config: %s", err)
+			}
+		})
+		if _, err = secretResolver.Resolve(yamlConf, origin); err != nil {
```

Comment:
Looking at the secretResolver functions, I find the names a bit cryptic.

`Subscribe` subscribe to what? When is the subscribe callback going to get called? What about changing the name to `RegisterResolveCallback` I think it express what do actual secret resolvers doing 😄

Reply:
I'm not a fan of `RegisterResolveCallback` (despite using it in an earlier version) because it's talking about a type that is usually not going to appear in code, since most users will be anonymous functions. I also think `ResolveCallback` needs an update because it's only mentioning `Resolve` but now secrets can be updated with `Refresh` as well. Going to go with `SecretChangeCallback` and `SubscribeToChanges` which I think together are a pretty good solution to all these issues.